## SWIS DATA ANALYSIS

In [47]:
import sys
sys.path.insert(0, '../')

import pandas as pd
import swis_data_info
import plotly.graph_objects as go

<img src="../images/WA.png">

In [4]:
swis_pc = swis_data_info.SWIS_postcodes

In [5]:
len(swis_pc)

142

In [6]:
swis_sites = swis_data_info.SWIS_site

In [7]:
len(swis_sites)

659

Read the data with time series information

In [10]:
data = pd.read_csv("../../all_ts_data/all_sites/data-statistics.csv", index_col = "site_index")

Extract the time series for the SWIS

In [17]:
site_ids_int = list(map(int, swis_sites))

In [19]:
swis_site_info = data.loc[data['site_id'].isin(site_ids_int)]

In [20]:
swis_site_info.head()

site_id                 Data Start                   Date End  \
site_index                                                                     
0           1948717897  2020-02-12 20:05:00+08:00  2021-03-10 16:00:00+08:00   
1            315046204  2020-11-09 00:00:00+08:00  2021-03-10 19:00:00+08:00   
2           1113196323  2020-02-12 20:05:00+08:00  2021-03-10 17:00:00+08:00   
3            838391054  2020-10-15 12:00:00+08:00  2021-03-10 19:00:00+08:00   
4             54265949  2020-02-12 20:05:00+08:00  2021-03-10 17:00:00+08:00   

            ts Length  Number of large gaps >=5hrs  \
site_index                                           
0              112848                            0   
1               35077                            1   
2              112860                            0   
3               42133                            0   
4              112860                            0   

            Number of small gaps <5hrs  \
site_index                               
0                                    0   
1                                    1   
2                                    0   
3                                    4   
4                                    0   

                                              Gap Start Dates  \
site_index                                                      
0                                                          []   
1           ['2020-11-15 22:10:00+08:00', '2020-11-16 00:0...   
2                                                          []   
3           ['2020-10-15 13:30:00+08:00', '2020-10-15 19:4...   
4                                                          []   

            Looks faulty (0 or 1 )  Curtailment \n(0 or 1)  \
site_index                                                   
0                                0                       1   
1                                0                       0   
2                                0                       1   
3                                0                       1   
4                                1                       1   

                                     Notes  
site_index                                  
0                                      NaN  
1                                      NaN  
2                                      NaN  
3                                      NaN  
4           * faulty - 0 values in between

In [21]:
len(swis_site_info)

659

We have 659 sites - Let's see the sites we can use. What are the conditions to use.

1) Faulty should be 0 - meaning the time series is not faulty

2) For now let's also take time series without large gap - so large gaps should be 0

3) We will also check whether time series has a length of at least 100,000

In [23]:
condition1 = swis_site_info.loc[swis_site_info['Looks faulty (0 or 1 )'] == 0]

In [24]:
len(condition1)

488

In [26]:
condition1.head()

site_id                 Data Start                   Date End  \
site_index                                                                     
0           1948717897  2020-02-12 20:05:00+08:00  2021-03-10 16:00:00+08:00   
1            315046204  2020-11-09 00:00:00+08:00  2021-03-10 19:00:00+08:00   
2           1113196323  2020-02-12 20:05:00+08:00  2021-03-10 17:00:00+08:00   
3            838391054  2020-10-15 12:00:00+08:00  2021-03-10 19:00:00+08:00   
7           1824904740  2020-10-26 14:15:00+08:00  2021-03-10 19:00:00+08:00   

            ts Length  Number of large gaps >=5hrs  \
site_index                                           
0              112848                            0   
1               35077                            1   
2              112860                            0   
3               42133                            0   
7               38938                            0   

            Number of small gaps <5hrs  \
site_index                               
0                                    0   
1                                    1   
2                                    0   
3                                    4   
7                                    2   

                                              Gap Start Dates  \
site_index                                                      
0                                                          []   
1           ['2020-11-15 22:10:00+08:00', '2020-11-16 00:0...   
2                                                          []   
3           ['2020-10-15 13:30:00+08:00', '2020-10-15 19:4...   
7           ['2020-10-26 17:05:00+08:00', '2020-11-08 23:2...   

            Looks faulty (0 or 1 )  Curtailment \n(0 or 1) Notes  
site_index                                                        
0                                0                       1   NaN  
1                                0                       0   NaN  
2                                0                       1   NaN  
3                                0                       1   NaN  
7                                0                       1   NaN

There are 488 time series that are not faulty. Now let's check from those which time series don't have large gaps

In [27]:
condition2 = condition1.loc[condition1['Number of large gaps >=5hrs'] == 0]

In [28]:
condition2.head()

site_id                 Data Start                   Date End  \
site_index                                                                     
0           1948717897  2020-02-12 20:05:00+08:00  2021-03-10 16:00:00+08:00   
2           1113196323  2020-02-12 20:05:00+08:00  2021-03-10 17:00:00+08:00   
3            838391054  2020-10-15 12:00:00+08:00  2021-03-10 19:00:00+08:00   
7           1824904740  2020-10-26 14:15:00+08:00  2021-03-10 19:00:00+08:00   
9            476839589  2020-06-25 18:20:00+08:00  2021-03-10 18:00:00+08:00   

            ts Length  Number of large gaps >=5hrs  \
site_index                                           
0              112848                            0   
2              112860                            0   
3               42133                            0   
7               38938                            0   
9               74301                            0   

            Number of small gaps <5hrs  \
site_index                               
0                                    0   
2                                    0   
3                                    4   
7                                    2   
9                                    0   

                                              Gap Start Dates  \
site_index                                                      
0                                                          []   
2                                                          []   
3           ['2020-10-15 13:30:00+08:00', '2020-10-15 19:4...   
7           ['2020-10-26 17:05:00+08:00', '2020-11-08 23:2...   
9                                                          []   

            Looks faulty (0 or 1 )  Curtailment \n(0 or 1) Notes  
site_index                                                        
0                                0                       1   NaN  
2                                0                       1   NaN  
3                                0                       1   NaN  
7                                0                       1   NaN  
9                                0                       1   NaN

In [29]:
len(condition2)

483

Now we have 483 time series which are not faulty and don't have large gaps in the data. Let's check their length now.

In [34]:
ts_swis = condition2.copy()

In [84]:
fig = go.Figure(data=[go.Histogram(x=ts_swis['ts Length'])])
fig.update_xaxes(title = 'Time series length')
fig.update_yaxes(title = 'Number of time series')
fig.show()

In [44]:
ts_length_info = pd.DataFrame(ts_swis.groupby(by = 'ts Length').size(), columns = ['Count']).reset_index()

In [46]:
ts_length_info.head()

ts Length  Count
0       1703      1
1       1791      1
2       2321      1
3       2345      1
4       2483      1

In [56]:
len(ts_length_info)

243

In [58]:
ts_length_info.max()

ts Length    112884
Count            77
dtype: int64

In [66]:
ts_swis_final = ts_swis.loc[ts_swis['ts Length'] >=100000]

In [67]:
ts_swis_final.head()

site_id                 Data Start                   Date End  \
site_index                                                                     
0           1948717897  2020-02-12 20:05:00+08:00  2021-03-10 16:00:00+08:00   
2           1113196323  2020-02-12 20:05:00+08:00  2021-03-10 17:00:00+08:00   
12          1597760039  2020-02-12 20:05:00+08:00  2021-03-10 15:00:00+08:00   
14           317532327  2020-02-12 20:05:00+08:00  2021-03-10 15:00:00+08:00   
16          1295864320  2020-02-12 20:05:00+08:00  2021-03-10 17:00:00+08:00   

            ts Length  Number of large gaps >=5hrs  \
site_index                                           
0              112848                            0   
2              112860                            0   
12             112836                            0   
14             112836                            0   
16             112860                            0   

            Number of small gaps <5hrs Gap Start Dates  \
site_index                                               
0                                    0              []   
2                                    0              []   
12                                   0              []   
14                                   0              []   
16                                   0              []   

            Looks faulty (0 or 1 )  Curtailment \n(0 or 1)  \
site_index                                                   
0                                0                       1   
2                                0                       1   
12                               0                       0   
14                               0                       1   
16                               0                       0   

                                                        Notes  
site_index                                                     
0                                                         NaN  
2                                                         NaN  
12                                        * May 6th seems off  
14          * drop after the peak looks weird - straight s...  
16                                                        NaN

In [68]:
len(ts_swis_final)

265

We have 265 time series. Now let's check how many postcodes these come from.

In [70]:
site_info = pd.read_csv("../../all_ts_data/all_sites/all_site_info.csv")

In [71]:
site_info.head()

site_id  dc_capacity  postcode
0  1354298257         6490      6123
1  1272386136        10540      6728
2  1942910158         6600      6014
3  1706523333        33550      6104
4  1068121303         3990      6330

In [75]:
swis_final_sites = ts_swis_final['site_id'].values.tolist()

In [77]:
swis_final_site_info = site_info.loc[site_info['site_id'].isin(swis_final_sites)]

In [79]:
swis_final_pcs = pd.DataFrame(swis_final_site_info.groupby('postcode').size(), columns = ['number of sites'])

In [82]:
swis_final_pcs

number of sites
postcode                 
6000                    1
6005                    1
6006                    2
6007                    1
6008                    1
...                   ...
6430                    2
6432                    2
6509                    1
6528                    1
6530                    8

[109 rows x 1 columns]

We have 109 postcodes in total for the final SWIS analysis